src/v/cloud_storage_clients/xml_sax_parser.cc

```diff
@@ -372,4 +373,51 @@ void xml_sax_parser::characters(
       {reinterpret_cast<const char*>(data), static_cast<size_t>(size)});
 }
 
+template<typename Impl>
+seastar::future<client::list_bucket_result> parse_from_stream(
+  seastar::input_stream<char> stream, std::optional<client::item_filter> pred) {
+    return ss::do_with(
+      std::move(stream),
+      xml_sax_parser{},
+      [pred = std::move(pred)](
+        ss::input_stream<char>& stream, xml_sax_parser& p) mutable {
+          p.start_parse(std::make_unique<Impl>(std::move(pred)));
+          return ss::do_until(
+                   [&stream] { return stream.eof(); },
+                   [&stream, &p] {
+                       return stream.read().then(
+                         [&p](ss::temporary_buffer<char> chunk) {
+                             /*
+                              * Seastar may return chunks of 128KB which is very
+                              * close to our max allocation limit. But when this
+                              * is handed off to to libxml the data may be
+                              * incorporated into a buffer that needs resizing
+                              * and exceed the max allocation limit. So break
+                              * this up into smaller chunks for processing.
+                              */
+                             constexpr auto max_part_size = 16_KiB;
+                             while (!chunk.empty()) {
+                                 auto part_size = std::min(
+                                   max_part_size, chunk.size());
+                                 p.parse_chunk(chunk.share(0, part_size));
+                                 chunk.trim_front(part_size);
+                             }
+                         });
+                   })
+            .then([&stream] { return stream.close(); })
```

Comment:
does this need to be finally? It's not clear

Reply:
i think it's probably good practice, but it's also quite rare for `close(2)` to fail so one might be interested in seeing it when it does happen. 🤷

Reply:
yes, that's good to do to avoid leaking open file descriptors. will fix that